### PR TITLE
[Merged by Bors] - feat(RingTheory/Congruence): add the `CompleteLattice` instance

### PR DESCRIPTION
**Pull request by eric-wieser**
The code is copied from `GroupTheory/Congruence.lean`, and then modified to fix the errors.

I haven't copied the full contents of `GroupTheory/Congruencee`, only the results about the lattice structure.

This replaces leanprover-community/mathlib#18588

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by YaelDillies — COMMENTED** [review]
Is there a specific reason you did not act on my comments on https://github.com/leanprover-community/mathlib/pull/18588 ? Would you mind doing so?

**Comment by eric-wieser** [comment]
Primarily because they requested cleanup that seemed out of scope (which is what the other PR i made is for). I'm happy to make this one depend on that one.

**Comment by YaelDillies** [comment]
I'm happy for the cleanups to happen later if you pinky-promise to actually do them.

File: Mathlib/RingTheory/Congruence.lean [hunk 1]
```diff
@@ -415,4 +419,161 @@ def mk' [NonAssocSemiring R] (c : RingCon R) : R →+* c.Quotient
...
+  le_top _ := fun _ _ _h => trivial
+  bot :=
+    { (⊥ : Setoid R) with
+      mul' := congr_arg₂ _
+      add' := congr_arg₂ _ }
+  bot_le c := fun x _y h => h ▸ c.refl x
+
+/-- The infimum of two congruence relations equals the infimum of the underlying binary
+operations. -/
+theorem coe_inf {c d : RingCon R} : ⇑(c ⊓ d) = ⇑c ⊓ ⇑d :=
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence.lean:495]:
```suggestion
@[simp, norm_cast]
theorem coe_inf {c d : RingCon R} : ⇑(c ⊓ d) = ⇑c ⊓ ⇑d :=
```
probably in other places too

**Review by YaelDillies — COMMENTED** [review]
What's the status wrt to the unresolved comments on the mathlib3 PR?

File: Mathlib/RingTheory/Congruence.lean [hunk 2]
```diff
@@ -415,4 +419,164 @@ def mk' [NonAssocSemiring R] (c : RingCon R) : R →+* c.Quotient
...
+instance : CompleteLattice (RingCon R) where
+  __ := completeLatticeOfInf (RingCon R) fun s =>
+    ⟨fun r hr x y h => (h : ∀ r ∈ s, (r : RingCon R) x y) r hr,
+      fun _r hr _x _y h _r' hr' => hr hr' h⟩
+  inf c d :=
+    { toSetoid := c.toSetoid ⊓ d.toSetoid
+      mul' := fun h1 h2 => ⟨c.mul h1.1 h2.1, d.mul h1.2 h2.2⟩
+      add' := fun h1 h2 => ⟨c.add h1.1 h2.1, d.add h1.2 h2.2⟩ }
+  inf_le_left _ _ := fun _ _ h => h.1
+  inf_le_right _ _ := fun _ _ h => h.2
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence.lean:480]: This should be obtained by pullback

File: Mathlib/RingTheory/Congruence.lean [hunk 3]
```diff
@@ -415,4 +419,164 @@ def mk' [NonAssocSemiring R] (c : RingCon R) : R →+* c.Quotient
...
+      add' := congr_arg₂ _ }
+  bot_le c := fun x _y h => h ▸ c.refl x
+
+@[simp, norm_cast]
+theorem coe_top : ⇑(⊤ : RingCon R) = ⊤ := rfl
+
+/-- The infimum of two congruence relations equals the infimum of the underlying binary
+operations. -/
+@[simp, norm_cast]
+theorem coe_inf {c d : RingCon R} : ⇑(c ⊓ d) = ⇑c ⊓ ⇑d := rfl
```
Comment (YaelDillies) [inline Mathlib/RingTheory/Congruence.lean:499]: Are you not missing `coe_sInf`, `coe_iInf`?

**Review by YaelDillies — APPROVED** [review]
Thanks!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**